flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicKeySelector.java

```diff
@@ -0,0 +1,384 @@
+                      + "should be included in equality fields: '%s'",
+                  partitionField,
+                  schema.columns().stream()
+                      .filter(c -> equalityFields.contains(c.name()))
+                      .collect(Collectors.toList()));
+            }
+            return partitionKeySelector(
+                tableName, schema, spec, writeParallelism, maxWriteParallelism);
+          }
+        }
+
+      case RANGE:
+        if (schema.identifierFieldIds().isEmpty()) {
+          LOG.warn(
+              "Fallback to use 'none' distribution mode, because there are no equality fields set "
+                  + "and {}=range is not supported yet in flink",
+              WRITE_DISTRIBUTION_MODE);
+          return tableKeySelector(tableName, writeParallelism, maxWriteParallelism);
+        } else {
+          LOG.info(
+              "Distribute rows by equality fields, because there are equality fields set "
+                  + "and{}=range is not supported yet in flink",
+              WRITE_DISTRIBUTION_MODE);
+          return equalityFieldKeySelector(
+              tableName, schema, equalityFields, writeParallelism, maxWriteParallelism);
+        }
+
+      default:
+        throw new IllegalArgumentException("Unrecognized " + WRITE_DISTRIBUTION_MODE + ": " + mode);
+    }
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> equalityFieldKeySelector(
+      String tableName,
+      Schema schema,
+      List<String> equalityFields,
+      int writeParallelism,
+      int maxWriteParallelism) {
+    return new TargetLimitedKeySelector(
+        new EqualityFieldKeySelector(
+            schema,
+            FlinkSchemaUtil.convert(schema),
+            DynamicRecordProcessor.getEqualityFieldIds(equalityFields, schema)),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> partitionKeySelector(
+      String tableName,
+      Schema schema,
+      PartitionSpec spec,
+      int writeParallelism,
+      int maxWriteParallelism) {
+    NonThrowingKeySelector<RowData, String> inner =
+        new PartitionKeySelector(spec, schema, FlinkSchemaUtil.convert(schema));
+    return new TargetLimitedKeySelector(
+        in -> inner.getKey(in).hashCode(),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> tableKeySelector(
+      String tableName, int writeParallelism, int maxWriteParallelism) {
+    return new TargetLimitedKeySelector(
+        new RoundRobinKeySelector<>(writeParallelism),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  /**
+   * Generates a new key using the salt as a base, and reduces the target key range of the {@link
+   * #wrapped} {@link NonThrowingKeySelector} to {@link #writeParallelism}.
```

Comment:
didn't quite get the purpose of this method. why would the `wrapped` use target key range higher than the `writeParallelism`. e.g. `RoundRobinKeySelector` was created with the same `writeParallelism as the outside `TargetLimitedKeySelector`

```
  private static NonThrowingKeySelector<RowData, Integer> tableKeySelector(
      String tableName, int writeParallelism, int maxWriteParallelism) {
    return new TargetLimitedKeySelector(
        new RoundRobinKeySelector<>(writeParallelism),
        tableName.hashCode(),
        writeParallelism,
        maxWriteParallelism);
  }
```

Reply:
The idea is that job parallelism and the write parallelism of each table are decoupled. For example, the job may run at parallelism 100, but a single table may run at write parallelism 10. In order to ensure the correct write parallelism, we generate keys such that they fall into the correct key groups for a subset of 10 tasks.

Note that there is an additional `salt` which will offset the range of subtasks, such that for every table we chose a different set of subtasks according to the table's writeParallelism.